src/server.ts

```diff
@@ -47,128 +39,180 @@ export interface ServerOptions {
+      request.on('message', this.onMessage(request, connectionSubscriptions));
+      request.on('close', () => {
+        this.onClose(request, connectionSubscriptions)();
+
+        if (this.onDisconnect) {
+          this.onDisconnect(request);
+        }
+      });
     });
   }
 
-  // TODO test that this actually works
-  private onClose(connection: Connection, connectionSubscriptions: ConnectionSubscriptions) {
+  private unsubscribe(connection: WebSocket, handleId: number) {
+    this.subscriptionManager.unsubscribe(handleId);
+
+    if (this.onUnsubscribe) {
+      this.onUnsubscribe(connection);
+    }
+  }
+
+  private onClose(connection: WebSocket, connectionSubscriptions: ConnectionSubscriptions) {
     return () => {
-      Object.keys(connectionSubscriptions).forEach( (subId) => {
-        this.subscriptionManager.unsubscribe(connectionSubscriptions[subId]);
+      Object.keys(connectionSubscriptions).forEach((subId) => {
+        this.unsubscribe(connection, connectionSubscriptions[subId]);
         delete connectionSubscriptions[subId];
       });
-    }
+    };
   }
 
-  private onMessage(connection: Connection, connectionSubscriptions: ConnectionSubscriptions, webSocketRequest: WebSocketRequest) {
-    return  (message: IMessage) => {
+  private onMessage(connection: WebSocket, connectionSubscriptions: ConnectionSubscriptions) {
+    return (message: any) => {
       let parsedMessage: SubscribeMessage;
       try {
-        parsedMessage = JSON.parse(message.utf8Data);
+        parsedMessage = JSON.parse(message);
       } catch (e) {
-        this.sendSubscriptionFail(connection, null, { errors: [{ message: e.message }] });
+        this.sendSubscriptionFail(connection, null, {errors: [{message: e.message}]});
         return;
       }
 
       const subId = parsedMessage.id;
       switch (parsedMessage.type) {
+        case INIT:
+          let onConnectPromise = Promise.resolve(true);
+          if (this.onConnect) {
+            onConnectPromise = Promise.resolve(this.onConnect(parsedMessage.payload, connection));
+          }
+
+          onConnectPromise.then((result) => {
+            if (result === false) {
+              throw new Error('Prohibited connection!');
+            }
+
+            this.initResult = result;
+
+            return {
+              type: INIT_SUCCESS,
+            };
+          }).catch((error: Error) => {
+            return {
+              type: INIT_FAIL,
+              error: error.message,
+            };
+          }).then((resultMessage: any) => {
+            this.sendInitResult(connection, resultMessage);
+          });
+
+          break;
 
         case SUBSCRIPTION_START:
           const baseParams: SubscriptionOptions = {
             query: parsedMessage.query,
             variables: parsedMessage.variables,
             operationName: parsedMessage.operationName,
-            context: {},
+            context: Object.assign({}, this.initResult),
```

Comment:
can't `this.initResult` be `true`?  Maaaybe that works with Object.assign but it sure feels weird.